### ISSUE TITLE
Training function

### ISSUE THREAD

**Issue by pushkarkadam** [pushkarkadam/handpose#13]
## Todo
- [x] Create a for loop for epoch
- [x] Change optimizer function to class
- [x] Change loss return type to a dictionary of all losses
- [x] Add saving network code
- [x] Wrap the code inside a train function.
- [x] Make a wrapper class for scheduler.
- [x] Add the mermaid flow chart to docs
- [x] Add loss graph
- [x] Fix the bug for epochs in the loss history
- [x] Fix eval dictionary
- [x] Plot mAP scores

## Flow chart

```mermaid
flowchart TD
    Z[CNN] -- outputs --> A[network_head] --pred_head --> B[head_activations]
    B -- pred_head_act --> C[best_box]
    C -- best_pred_head --> D[extract_head]
    D -- pred_data --> E[non_max_suppression]
    D -- pred_data --> F[mean_average_precision]
    E -- pred_data_nms --> F
```

**Comment by pushkarkadam** [comment]
The multiple losses are due to the batch sizes that are being iterated. Needs to be investigated.

**Comment by pushkarkadam** [comment]
Find a way to sum up `running_;loss` to compute `epoch_loss` for each of the types of losses in each batch. Then, add the epoch losses to the dictionary of `all_losses`

**Comment by pushkarkadam** [comment]
Maybe include a variable `running_total_loss`, `running_box_loss`, ... such that they are computed in each epoch and then after the epoch, they are appended to the `all_losses` dictionary.

**Comment by pushkarkadam** [comment]
In the training on dev dataset, mAP gives 0.0 for all the train and valid epochs. Though, the unit tests are running correctly, and this could be the issue due to small train and valid datasize.

I will commit the training code in its present state as it is completing the required functions.
